File: generated_code/GeminiPro/Java/020_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-adjusting binary search tree with the additional property
 * that recently accessed elements are quick to access again. All major operations
 * (insertion, deletion, search) involve a "splaying" step, which moves the accessed
 * node to the root of the tree.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node parent;
        Node left;
        Node right;

        /**
         * Constructs a new node with the given key.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.parent = null;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a left rotation on the given node x.
     * This operation is a fundamental building block for the splay operation.
     * @param x The node to rotate.
     */
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y != null) {
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.left = x;
            x.parent = y;
        }
    }

    /**
     * Performs a right rotation on the given node y.
     * This operation is a fundamental building block for the splay operation.
     * @param y The node to rotate.
     */
    private void rightRotate(Node y) {
        Node x = y.left;
        if (x != null) {
            y.left = x.right;
            if (x.right != null) {
                x.right.parent = y;
            }
            x.parent = y.parent;
            if (y.parent == null) {
                this.root = x;
            } else if (y == y.parent.right) {
                y.parent.right = x;
            } else {
                y.parent.left = x;
            }
            x.right = y;
            y.parent = x;
        }
    }

    /**
     * Performs the splay operation on a node x, moving it to the root of the tree
     * through a series of rotations.
     * @param x The node to splay.
     */
    private void splay(Node x) {
        while (x.parent != null) {
            Node parent = x.parent;
            Node grandParent = parent.parent;
            if (grandParent == null) { // Zig step
                if (x == parent.left) {
                    rightRotate(parent);
                } else {
                    leftRotate(parent);
                }
            } else {
                if (x == parent.left) {
                    if (parent == grandParent.left) { // Zig-Zig step
                        rightRotate(grandParent);
                        rightRotate(parent);
                    } else { // Zig-Zag step
                        rightRotate(parent);
                        leftRotate(grandParent);
                    }
                } else {
                    if (parent == grandParent.right) { // Zig-Zig step
                        leftRotate(grandParent);
                        leftRotate(parent);
                    } else { // Zig-Zag step
                        leftRotate(parent);
                        rightRotate(grandParent);
                    }
                }
            }
        }
    }

    /**
     * Searches for a key in the splay tree.
     * If the key is found, the corresponding node is splayed to the root.
     * If the key is not found, the last non-null node on the search path is splayed to the root.
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }

        Node current = root;
        Node lastNode = root;

        while (current != null) {
            lastNode = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key found
                splay(current);
                return current.key;
            }
        }

        // Key not found, splay the last visited node
        splay(lastNode);
        
        return null;
    }
    
    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the node with that key is splayed to the root.
     * Otherwise, a new node is created, inserted, and then splayed to the root.
     * @param key The key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        Node current = root;
        Node parent = null;
        while (current != null) {
            parent = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key already exists, splay the node and return as it's a set
                splay(current);
                return;
            }
        }

        Node newNode = new Node(key);
        newNode.parent = parent;
        if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }
        splay(newNode);
    }

    /**
     * Deletes a key from the splay tree.
     * The tree is first searched for the key, which brings the node (or the closest node)
     * to the root. If the key is found at the root, it is removed, and its subtrees
     * are joined.
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // search() will splay the found node or the last accessed node to the root
        search(key);

        // If the key is not at the root after searching, it wasn't in the tree
        if (root.key != key) {
            return;
        }

        // The node to delete is now the root
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
            if (root != null) {
                root.parent = null;
            }
        } else {
            // Detach the left subtree to operate on it
            leftSubtree.parent = null;

            // Find the maximum node in the left subtree
            Node maxNode = leftSubtree;
            while (maxNode.right != null) {
                maxNode = maxNode.right;
            }

            // Temporarily make the left subtree the main tree to splay within it
            this.root = leftSubtree;
            splay(maxNode); // maxNode becomes the new root of this subtree

            // Attach the original right subtree
            root.right = rightSubtree;
            if (rightSubtree != null) {
                rightSubtree.parent = root;
            }
        }
    }
}

File: generated_code/GeminiPro/Python/020_SplayTree.py
class SplayTree:
    """
    A complete, self-contained Python class that implements a dictionary-like set
    for integers using a Splay Tree.

    The main operations are insert, delete, and search. Any access to a node
    (or a would-be node) causes a "splay" operation, which moves the accessed
    node (or its parent) to the root of the tree. This restructuring helps to
    maintain an amortized O(log n) time complexity for all operations.
    """

    class Node:
        """
        An inner class representing a node in the Splay Tree.
        Each node stores a key and pointers to its left child, right child,
        and parent, which are essential for the rotation and splaying logic.
        """
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """
        Performs a left rotation on the given node x.
        This operation is a fundamental building block for splaying.
        """
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """
        Performs a right rotation on the given node x.
        This operation is a fundamental building block for splaying.
        """
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Performs the splay operation on a node, moving it to the root
        through a series of rotations. The type of rotation (Zig, Zig-Zig,
        or Zig-Zag) depends on the configuration of the node, its parent,
        and its grandparent.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-Zig case
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zig-Zig case
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:  # Zig-Zag case
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag case (node is left child, parent is right child)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree and performs the splaying operation.

        If the key is found, the corresponding node is splayed to the root, and
        the method returns True. If the key is not found, the last non-null node
        visited during the search (the would-be parent of the key) is splayed
        to the root, and the method returns False.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_visited = None
        while current:
            last_visited = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # Key not found, splay the last visited node to the root
        if last_visited:
            self._splay(last_visited)
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        This method follows the standard bottom-up insertion for Splay Trees.
        It first performs a regular Binary Search Tree insertion. Then, the
        newly inserted node is splayed to the root. If the key already exists,
        the existing node is splayed to the root and no new node is inserted.

        Args:
            key (int): The integer key to insert.
        """
        # Find parent for new node using standard BST logic
        parent = None
        current = self.root
        while current:
            parent = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay the node and return
                self._splay(current)
                return

        # Create and link the new node
        new_node = self.Node(key)
        new_node.parent = parent
        if not parent:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the newly inserted node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        First, the key is searched for, which brings the node (or its would-be
        parent) to the root. If the key is found at the root, the tree is split
        into two subtrees (left and right). The maximum element of the left
        subtree is then found, splayed to the root of the left subtree, and
        the right subtree is attached as its right child.

        Args:
            key (int): The integer key to delete.
        """
        # Search for the key. This splays the node (if found) or its
        # would-be parent to the root.
        if not self.search(key):
            # Key not in the tree. search() has already splayed the last
            # visited node. Nothing more to do.
            return

        # If search returned true, the node to delete is now at the root.
        z = self.root
        left_subtree = z.left
        right_subtree = z.right

        if not left_subtree:
            # No left child, so the right subtree becomes the new tree.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # The left subtree exists. We will make the maximum element of the
            # left subtree the new root of the combined tree.
            
            # Isolate the left subtree to perform a splay within it.
            left_subtree.parent = None
            
            # Find the max node in the now-isolated left subtree.
            new_root_candidate = left_subtree
            while new_root_candidate.right:
                new_root_candidate = new_root_candidate.right
            
            # Splay the max node to the top of its subtree.
            # We can do this by temporarily making that subtree the main tree.
            self.root = left_subtree
            self._splay(new_root_candidate)
            
            # After splaying, `self.root` is the root of the splayed left
            # subtree (the max node). We re-attach the original right subtree.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root